File: internal/deps/deps.go
// Package deps answers "which dependencies did AI-written code pull in, and do
// they even exist?" Agents hallucinate package names; attackers register
// them (slopsquatting). grain already knows which added lines an agent wrote,
// so it can list the dependencies those lines introduced, whether anyone
// reviewed them, and, when asked to consult the registries, whether each
// package exists and how old it is. Signals, not verdicts.
package deps

import (
	"context"
	"encoding/json"
	"net/http"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/FrontTribe/grain/internal/config"
	"github.com/FrontTribe/grain/internal/gitlog"
	"github.com/FrontTribe/grain/internal/outcomes"
	"github.com/FrontTribe/grain/internal/risk"
	"github.com/FrontTribe/grain/internal/signal"
)

// Dep is one dependency an added line introduced.
type Dep struct {
	Name      string `json:"name"`
	Ecosystem string `json:"ecosystem"` // npm | go | pypi | cargo | rubygems
	Manifest  string `json:"manifest"`
	SHA       string `json:"sha"`
	Subject   string `json:"subject"`
	AI        bool   `json:"ai"`
	Reviewed  bool   `json:"reviewed"`
	// Registry answer, when checked: exists, and age in days at scan time.
	Checked bool   `json:"checked"`
	Exists  bool   `json:"exists"`
	AgeDays int    `json:"age_days"` // -1 when unknown
	URL     string `json:"url"`
}

// Summary is the dependencies block of a report.
type Summary struct {
	Total        int    `json:"total"`
	AI           int    `json:"ai"`
	AIUnreviewed int    `json:"ai_unreviewed"`
	Checked      bool   `json:"checked"` // whether registries were consulted
	Missing      int    `json:"missing"` // checked and not found
	Young        int    `json:"young"`   // checked and younger than YoungDays
	Deps         []Dep  `json:"deps"`    // worst first, capped
	Commits      int    `json:"commits"`
	Source       string `json:"source"`
}

// YoungDays is the age under which a package counts as "young": typical of a
// name registered after an agent started hallucinating it.
const YoungDays = 30

const maxDeps = 60

var manifestRe = regexp.MustCompile(`(?i)(^|/)(package\.json|go\.mod|requirements[\w.-]*\.txt|pyproject\.toml|Cargo\.toml|Gemfile)$`)

// IsManifest reports whether a path is a dependency manifest grain reads.
func IsManifest(path string) bool { return manifestRe.MatchString(path) }

var (
	// package.json: "name": "<version range>" lines. Top-level fields and
	// engines are excluded by key; scripts by value shape.
	npmLine   = regexp.MustCompile(`^\s*"(@?[a-z0-9][a-z0-9._-]*(?:/[a-z0-9._-]+)?)"\s*:\s*"([^"]*)"\s*,?\s*$`)
	npmRange  = regexp.MustCompile(`^(\^|~|>=?|<=?|=|\*$|latest$|next$|\d|workspace:|npm:|file:|link:|git\+|github:|https?://)`)
	npmNotDep = map[string]bool{
		"name": true, "version": true, "description": true, "main": true, "module": true, "types": true, "typings": true,
		"license": true, "author": true, "homepage": true, "type": true, "packageManager": true, "node": true, "npm": true,
		"yarn": true, "pnpm": true, "bun": true, "engine-strict": true, "browser": true, "exports": true, "files": true,
		"url": true, "bugs": true, "funding": true, "repository": true, "keywords": true, "private": true, "sideEffects": true,
		"publishConfig": true, "bin": true, "directories": true, "os": true, "cpu": true, "email": true, "scripts": true,
	}
	goLine   = regexp.MustCompile(`^\s*([a-z0-9][a-z0-9.-]*\.[a-z]{2,}/[\w./~-]+)\s+v\d[\w.+-]*\s*$`)
	pyReqRe  = regexp.MustCompile(`^\s*([A-Za-z0-9][A-Za-z0-9._-]*)\s*(\[[^\]]*\])?\s*(==|>=|<=|~=|!=|>|<|===|;|$)`)
	pyArrRe  = regexp.MustCompile(`^\s*"([A-Za-z0-9][A-Za-z0-9._-]*)(\[[^\]]*\])?\s*[^"]*",?\s*$`)
	poetryRe = regexp.MustCompile(`^\s*([A-Za-z0-9][A-Za-z0-9._-]*)\s*=\s*("[^"]*"|\{)`)
	cargoRe  = regexp.MustCompile(`^\s*([A-Za-z0-9][A-Za-z0-9_-]*)\s*=\s*("[^"]*"|\{)`)
	cargoNot = map[string]bool{"name": true, "version": true, "edition": true, "authors": true, "description": true, "license": true,
		"repository": true, "readme": true, "keywords": true, "categories": true, "build": true, "rust-version": true, "default-run": true,
		"publish": true, "homepage": true, "documentation": true, "exclude": true, "include": true, "workspace": true, "resolver": true,
		"members": true, "path": true, "features": true, "default": true, "opt-level": true, "lto": true, "codegen-units": true, "panic": true}
	gemRe = regexp.MustCompile(`^\s*gem\s+['"]([^'"]+)['"]`)
	pyNot = map[string]bool{"python": true, "pip": true, "setuptools": true, "wheel": true,
		"name": true, "version": true, "description": true, "readme": true, "license": true, "authors": true, "maintainers": true,
		"requires-python": true, "dependencies": true, "keywords": true, "classifiers": true, "urls": true, "homepage": true, "repository": true, "documentation": true, "packages": true, "include": true, "exclude": true}
)

// Parse extracts the dependency names an added line of a manifest introduces.
func Parse(path, line string) (name, ecosystem string, ok bool) {
	base := path
	if i := strings.LastIndexByte(path, '/'); i >= 0 {
		base = path[i+1:]
	}
	lower := strings.ToLower(base)
	t := strings.TrimSpace(line)
	if t == "" || strings.HasPrefix(t, "#") || strings.HasPrefix(t, "//") {
		return "", "", false
	}
	switch {
	case lower == "package.json":
		m := npmLine.FindStringSubmatch(line)
		if m == nil || npmNotDep[m[1]] || !npmRange.MatchString(m[2]) {
			return "", "", false
		}
		return m[1], "npm", true
	case lower == "go.mod":
		if strings.Contains(line, "// indirect") {
			return "", "", false
		}
		m := goLine.FindStringSubmatch(line)
		if m == nil {
			return "", "", false
		}
		return m[1], "go", true
	case strings.HasPrefix(lower, "requirements") && strings.HasSuffix(lower, ".txt"):
		if strings.HasPrefix(t, "-") || strings.Contains(t, "://") {
			return "", "", false
		}
		m := pyReqRe.FindStringSubmatch(t)
		if m == nil || pyNot[strings.ToLower(m[1])] {
			return "", "", false
		}
		return strings.ToLower(m[1]), "pypi", true
	case lower == "pyproject.toml":
		if m := pyArrRe.FindStringSubmatch(line); m != nil && !pyNot[strings.ToLower(m[1])] {
			return strings.ToLower(m[1]), "pypi", true
		}
		if m := poetryRe.FindStringSubmatch(line); m != nil && !pyNot[strings.ToLower(m[1])] && !strings.Contains(m[1], ".") {
			return strings.ToLower(m[1]), "pypi", true
		}
		return "", "", false
	case lower == "cargo.toml":
		m := cargoRe.FindStringSubmatch(line)
		if m == nil || cargoNot[m[1]] {
			return "", "", false
		}
		return m[1], "cargo", true
	case lower == "gemfile":
		m := gemRe.FindStringSubmatch(line)
		if m == nil {
			return "", "", false
		}
		return m[1], "rubygems", true
	}
	return "", "", false
}

// Lookup asks a public registry whether a package exists and when it was
// created. Only the package name leaves the machine. ok=false when the
// registry couldn't be reached or answered unexpectedly.
func Lookup(ctx context.Context, ecosystem, name string) (exists bool, ageDays int, ok bool) {
	client := &http.Client{Timeout: 6 * time.Second}
	get := func(u string) (int, map[string]any) {
		req, _ := http.NewRequestWithContext(ctx, "GET", u, nil)
		req.Header.Set("User-Agent", "grain (https://github.com/FrontTribe/grain)")
		req.Header.Set("Accept", "application/json")
		res, err := client.Do(req)
		if err != nil {
			return 0, nil
		}
		defer res.Body.Close()
		var body map[string]any
		_ = json.NewDecoder(res.Body).Decode(&body)
		return res.StatusCode, body
	}
	age := func(iso string) int {
		t, err := time.Parse(time.RFC3339Nano, iso)
		if err != nil {
			if t, err = time.Parse("2006-01-02T15:04:05", iso); err != nil {
				return -1
			}
		}
		return int(time.Since(t).Hours() / 24)
	}
	switch ecosystem {
	case "npm":
		code, body := get("https://registry.npmjs.org/" + url.PathEscape(name))
		if code == 404 {
			return false, -1, true
		}
		if code != 200 {
			return false, -1, false
		}
		if tm, _ := body["time"].(map[string]any); tm != nil {
			if c, _ := tm["created"].(string); c != "" {
				return true, age(c), true
			}
		}
		return true, -1, true
	case "pypi":
		code, body := get("https://pypi.org/pypi/" + url.PathEscape(name) + "/json")
		if code == 404 {
			return false, -1, true
		}
		if code != 200 {
			return false, -1, false
		}
		// Age: the earliest upload across releases.
		earliest := -1
		if rel, _ := body["releases"].(map[string]any); rel != nil {
			for _, files := range rel {
				for _, f := range asSlice(files) {
					if fm, _ := f.(map[string]any); fm != nil {
						if up, _ := fm["upload_time_iso_8601"].(string); up != "" {
							if a := age(up); a > earliest {
								earliest = a
							}
						}
					}
				}
			}
		}
		return true, earliest, true
	case "go":
		code, _ := get("https://proxy.golang.org/" + escapeGoModule(name) + "/@latest")
		if code == 404 || code == 410 {
			return false, -1, true
		}
		if code != 200 {
			return false, -1, false
		}
		return true, -1, true
	case "cargo":
		code, body := get("https://crates.io/api/v1/crates/" + url.PathEscape(name))
		if code == 404 {
			return false, -1, true
		}
		if code != 200 {
			return false, -1, false
		}
		if c, _ := body["crate"].(map[string]any); c != nil {
			if created, _ := c["created_at"].(string); created != "" {
				return true, age(created), true
			}
		}
		return true, -1, true
	case "rubygems":
		code, body := get("https://rubygems.org/api/v1/gems/" + url.PathEscape(name) + ".json")
		if code == 404 {
			return false, -1, true
		}
		if code != 200 {
			return false, -1, false
		}
		if created, _ := body["created_at"].(string); created != "" {
			return true, age(created), true
		}
		return true, -1, true
	}
	return false, -1, false
}

func asSlice(v any) []any {
	s, _ := v.([]any)
	return s
}

// escapeGoModule applies the Go proxy's case encoding (upper → !lower).
func escapeGoModule(m string) string {
	var b strings.Builder
	for _, r := range m {
		if r >= 'A' && r <= 'Z' {
			b.WriteByte('!')
			b.WriteRune(r + ('a' - 'A'))
		} else {
			b.WriteRune(r)
		}
	}
	return b.String()
}

// RegistryURL is where a human would look the package up.
func RegistryURL(ecosystem, name string) string {
	switch ecosystem {
	case "npm":
		return "https://www.npmjs.com/package/" + name
	case "pypi":
		return "https://pypi.org/project/" + name + "/"
	case "go":
		return "https://pkg.go.dev/" + name
	case "cargo":
		return "https://crates.io/crates/" + name
	case "rubygems":
		return "https://rubygems.org/gems/" + name
	}
	return ""
}

// Compute lists the dependencies added over the commits, attributed like
// risk.Compute (attested line hashes, else declared commits), and optionally
// checks each distinct name against its registry.
func Compute(ctx context.Context, commits []gitlog.Commit, added map[string]map[string][]string, firstParent map[string]bool, cfg config.Config, checkRegistry bool) Summary {
	sum := Summary{Commits: len(commits), Source: "cli", Checked: checkRegistry}
	seen := map[string]bool{} // ecosystem/name: first introduction wins
	var deps []Dep

	// Oldest first, so a dependency is attributed to the commit that added it.
	for i := len(commits) - 1; i >= 0; i-- {
		c := commits[i]
		files := added[c.SHA]
		if len(files) == 0 {
			continue
		}
		var s signal.Set
		var hashes map[string]bool
		attributed := false
		for path, lines := range files {
			if !IsManifest(path) {
				continue
			}
			if !attributed {
				s = signal.Extract(c, cfg)
				hashes = outcomes.AIHashes(c.Note)
				attributed = true
			}
			lineLevel := len(hashes) > 0
			wholeAI := !lineLevel && (s.AttestedClass == "ai" || s.DeclaredAI)
			reviewed := risk.Reviewed(c, firstParent)
			for _, l := range lines {
				name, eco, ok := Parse(path, l)
				if !ok {
					continue
				}
				key := eco + "/" + name
				if seen[key] {
					continue
				}
				seen[key] = true
				ai := wholeAI || (lineLevel && outcomes.Substantive(l) && hashes[outcomes.LineHash(l)])
				deps = append(deps, Dep{
					Name: name, Ecosystem: eco, Manifest: path, SHA: c.SHA, Subject: strings.TrimSpace(c.Subject),
					AI: ai, Reviewed: reviewed, AgeDays: -1, URL: RegistryURL(eco, name),
				})
			}
		}
	}

	if checkRegistry {
		for i := range deps {
			d := &deps[i]
			exists, ageDays, ok := Lookup(ctx, d.Ecosystem, d.Name)
			if !ok {
				continue
			}
			d.Checked = true
			d.Exists = exists
			d.AgeDays = ageDays
		}
	}

	for _, d := range deps {
		sum.Total++
		if d.AI {
			sum.AI++
			if !d.Reviewed {
				sum.AIUnreviewed++
			}
		}
		if d.Checked && !d.Exists {
			sum.Missing++
		}
		if d.Checked && d.Exists && d.AgeDays >= 0 && d.AgeDays < YoungDays {
			sum.Young++
		}
	}
	sort.SliceStable(deps, func(i, j int) bool { return rank(deps[i]) > rank(deps[j]) })
	if len(deps) > maxDeps {
		deps = deps[:maxDeps]
	}
	sum.Deps = deps
	return sum
}

func rank(d Dep) int {
	r := 0
	if d.Checked && !d.Exists {
		r += 16
	}
	if d.Checked && d.Exists && d.AgeDays >= 0 && d.AgeDays < YoungDays {
		r += 8
	}
	if d.AI {
		r += 4
	}
	if !d.Reviewed {
		r += 2
	}
	return r
}

File: web/src/lib/deps.ts
// Cloud-side dependency analysis: which dependencies the scanned commits
// added, attributed like the risk pass (AI-written or not, review evidence or
// not), and checked against the public registries: does the package exist,
// and how old is it. Agents hallucinate names; attackers register them
// (slopsquatting). Mirrors internal/deps; only package names leave the server.
import { lineHash, substantive, type CloudRiskCommit, type ReviewEvidence } from "@/lib/risk";
import { triageDeps, suspicious, type DepTriage } from "@/lib/triage";

export type Ecosystem = "npm" | "go" | "pypi" | "cargo" | "rubygems";

export type CloudDep = {
  name: string;
  ecosystem: Ecosystem;
  manifest: string;
  sha: string;
  subject: string;
  ai: boolean;
  reviewed: boolean;
  checked: boolean;
  exists: boolean;
  age_days: number; // -1 when unknown
  url: string;
  triage?: DepTriage; // Cloud name triage (TypeSafe Jev), when enabled
};

export type CloudDeps = {
  total: number;
  ai: number;
  ai_unreviewed: number;
  checked: boolean;
  missing: number;
  young: number;
  suspicious: number; // names that look typosquatted or invented (triage)
  deps: CloudDep[];
  commits: number;
  source: "cloud";
};

export const YOUNG_DAYS = 30;
const MAX_DEPS = 60;

const manifestRe = /(^|\/)(package\.json|go\.mod|requirements[\w.-]*\.txt|pyproject\.toml|Cargo\.toml|Gemfile)$/i;
export function isManifest(path: string): boolean {
  return manifestRe.test(path);
}

const npmLine = /^\s*"(@?[a-z0-9][a-z0-9._-]*(?:\/[a-z0-9._-]+)?)"\s*:\s*"([^"]*)"\s*,?\s*$/;
const npmRange = /^(\^|~|>=?|<=?|=|\*$|latest$|next$|\d|workspace:|npm:|file:|link:|git\+|github:|https?:\/\/)/;
const npmNotDep = new Set(["name", "version", "description", "main", "module", "types", "typings", "license", "author", "homepage", "type", "packageManager", "node", "npm", "yarn", "pnpm", "bun", "engine-strict", "browser", "exports", "files", "url", "bugs", "funding", "repository", "keywords", "private", "sideEffects", "publishConfig", "bin", "directories", "os", "cpu", "email", "scripts"]);
const goLine = /^\s*([a-z0-9][a-z0-9.-]*\.[a-z]{2,}\/[\w./~-]+)\s+v\d[\w.+-]*\s*$/;
const pyReqRe = /^\s*([A-Za-z0-9][A-Za-z0-9._-]*)\s*(\[[^\]]*\])?\s*(==|>=|<=|~=|!=|>|<|===|;|$)/;
const pyArrRe = /^\s*"([A-Za-z0-9][A-Za-z0-9._-]*)(\[[^\]]*\])?\s*[^"]*",?\s*$/;
const poetryRe = /^\s*([A-Za-z0-9][A-Za-z0-9._-]*)\s*=\s*("[^"]*"|\{)/;
const cargoRe = /^\s*([A-Za-z0-9][A-Za-z0-9_-]*)\s*=\s*("[^"]*"|\{)/;
const cargoNot = new Set(["name", "version", "edition", "authors", "description", "license", "repository", "readme", "keywords", "categories", "build", "rust-version", "default-run", "publish", "homepage", "documentation", "exclude", "include", "workspace", "resolver", "members", "path", "features", "default", "opt-level", "lto", "codegen-units", "panic"]);
const gemRe = /^\s*gem\s+['"]([^'"]+)['"]/;
const pyNot = new Set(["python", "pip", "setuptools", "wheel", "name", "version", "description", "readme", "license", "authors", "maintainers", "requires-python", "dependencies", "keywords", "classifiers", "urls", "homepage", "repository", "documentation", "packages", "include", "exclude"]);

// The dependency an added manifest line introduces, if any.
export function parseDep(path: string, line: string): { name: string; ecosystem: Ecosystem } | null {
  const base = path.slice(path.lastIndexOf("/") + 1).toLowerCase();
  const t = line.trim();
  if (!t || t.startsWith("#") || t.startsWith("//")) return null;
  if (base === "package.json") {
    const m = npmLine.exec(line);
    if (!m || npmNotDep.has(m[1]) || !npmRange.test(m[2])) return null;
    return { name: m[1], ecosystem: "npm" };
  }
  if (base === "go.mod") {
    if (line.includes("// indirect")) return null;
    const m = goLine.exec(line);
    return m ? { name: m[1], ecosystem: "go" } : null;
  }
  if (base.startsWith("requirements") && base.endsWith(".txt")) {
    if (t.startsWith("-") || t.includes("://")) return null;
    const m = pyReqRe.exec(t);
    if (!m || pyNot.has(m[1].toLowerCase())) return null;
    return { name: m[1].toLowerCase(), ecosystem: "pypi" };
  }
  if (base === "pyproject.toml") {
    const a = pyArrRe.exec(line);
    if (a && !pyNot.has(a[1].toLowerCase())) return { name: a[1].toLowerCase(), ecosystem: "pypi" };
    const p = poetryRe.exec(line);
    if (p && !pyNot.has(p[1].toLowerCase()) && !p[1].includes(".")) return { name: p[1].toLowerCase(), ecosystem: "pypi" };
    return null;
  }
  if (base === "cargo.toml") {
    const m = cargoRe.exec(line);
    if (!m || cargoNot.has(m[1])) return null;
    return { name: m[1], ecosystem: "cargo" };
  }
  if (base === "gemfile") {
    const m = gemRe.exec(line);
    return m ? { name: m[1], ecosystem: "rubygems" } : null;
  }
  return null;
}

export function registryUrl(eco: Ecosystem, name: string): string {
  switch (eco) {
    case "npm": return `https://www.npmjs.com/package/${name}`;
    case "pypi": return `https://pypi.org/project/${name}/`;
    case "go": return `https://pkg.go.dev/${name}`;
    case "cargo": return `https://crates.io/crates/${name}`;
    case "rubygems": return `https://rubygems.org/gems/${name}`;
  }
}

const ageDays = (iso: string): number => {
  const t = Date.parse(iso);
  return Number.isFinite(t) ? Math.floor((Date.now() - t) / 86_400_000) : -1;
};

const escapeGoModule = (m: string) => m.replace(/[A-Z]/g, (c) => "!" + c.toLowerCase());

// Ask a registry whether a package exists and when it was created. Returns
// null when the registry couldn't be reached or answered unexpectedly.
export async function lookup(eco: Ecosystem, name: string): Promise<{ exists: boolean; age_days: number } | null> {
  const headers = { "User-Agent": "grain (https://github.com/FrontTribe/grain)", Accept: "application/json" };
  const get = async (url: string): Promise<{ status: number; body: unknown }> => {
    try {
      const res = await fetch(url, { headers, cache: "no-store", signal: AbortSignal.timeout(6000) });
      let body: unknown = null;
      try { body = await res.json(); } catch { body = null; }
      return { status: res.status, body };
    } catch {
      return { status: 0, body: null };
    }
  };
  type Obj = Record<string, unknown>;
  switch (eco) {
    case "npm": {
      const r = await get(`https://registry.npmjs.org/${encodeURIComponent(name)}`);
      if (r.status === 404) return { exists: false, age_days: -1 };
      if (r.status !== 200) return null;
      const created = ((r.body as Obj)?.time as Obj | undefined)?.created;
      return { exists: true, age_days: typeof created === "string" ? ageDays(created) : -1 };
    }
    case "pypi": {
      const r = await get(`https://pypi.org/pypi/${encodeURIComponent(name)}/json`);
      if (r.status === 404) return { exists: false, age_days: -1 };
      if (r.status !== 200) return null;
      let earliest = -1;
      const releases = (r.body as Obj)?.releases as Record<string, Obj[]> | undefined;
      for (const files of Object.values(releases ?? {})) {
        for (const f of files ?? []) {
          const up = f?.upload_time_iso_8601;
          if (typeof up === "string") earliest = Math.max(earliest, ageDays(up));
        }
      }
      return { exists: true, age_days: earliest };
    }
    case "go": {
      const r = await get(`https://proxy.golang.org/${escapeGoModule(name)}/@latest`);
      if (r.status === 404 || r.status === 410) return { exists: false, age_days: -1 };
      if (r.status !== 200) return null;
      return { exists: true, age_days: -1 };
    }
    case "cargo": {
      const r = await get(`https://crates.io/api/v1/crates/${encodeURIComponent(name)}`);
      if (r.status === 404) return { exists: false, age_days: -1 };
      if (r.status !== 200) return null;
      const created = ((r.body as Obj)?.crate as Obj | undefined)?.created_at;
      return { exists: true, age_days: typeof created === "string" ? ageDays(created) : -1 };
    }
    case "rubygems": {
      const r = await get(`https://rubygems.org/api/v1/gems/${encodeURIComponent(name)}.json`);
      if (r.status === 404) return { exists: false, age_days: -1 };
      if (r.status !== 200) return null;
      const created = (r.body as Obj)?.created_at;
      return { exists: true, age_days: typeof created === "string" ? ageDays(created) : -1 };
    }
  }
}

function rank(d: CloudDep): number {
  return (suspicious(d.triage, d) ? 32 : 0) + (d.checked && !d.exists ? 16 : 0) + (d.checked && d.exists && d.age_days >= 0 && d.age_days < YOUNG_DAYS ? 8 : 0) + (d.ai ? 4 : 0) + (d.reviewed ? 0 : 2);
}

// Commits arrive newest first; walk them oldest first so a dependency is
// attributed to the commit that introduced it.
export async function computeCloudDeps(opts: { commits: CloudRiskCommit[]; evidence: Map<string, ReviewEvidence>; checkRegistry?: boolean }): Promise<CloudDeps> {
  const check = opts.checkRegistry ?? true;
  const seen = new Set<string>();
  const deps: CloudDep[] = [];
  for (let i = opts.commits.length - 1; i >= 0; i--) {
    const c = opts.commits[i];
    if (!c.added) continue;
    const reviewed = (opts.evidence.get(c.sha) ?? "none") !== "none";
    for (const [path, lines] of Object.entries(c.added)) {
      if (!isManifest(path)) continue;
      for (const l of lines) {
        const d = parseDep(path, l);
        if (!d) continue;
        const key = `${d.ecosystem}/${d.name}`;
        if (seen.has(key)) continue;
        seen.add(key);
        const ai = c.wholeAI || (!!c.aiHashes && substantive(l) && c.aiHashes.has(lineHash(l)));
        deps.push({ name: d.name, ecosystem: d.ecosystem, manifest: path, sha: c.sha, subject: (c.message.split("\n")[0] ?? "").trim(), ai, reviewed, checked: false, exists: false, age_days: -1, url: registryUrl(d.ecosystem, d.name) });
      }
    }
  }
  if (check) {
    await Promise.all(deps.map(async (d) => {
      const r = await lookup(d.ecosystem, d.name);
      if (r) { d.checked = true; d.exists = r.exists; d.age_days = r.age_days; }
    }));
  }
  // Name triage after the registry answer, so the model sees it too.
  const triage = await triageDeps(deps);
  for (const d of deps) {
    const t = triage.get(`${d.ecosystem}/${d.name}`);
    if (t) d.triage = t;
  }
  const out: CloudDeps = { total: deps.length, ai: 0, ai_unreviewed: 0, checked: check, missing: 0, young: 0, suspicious: 0, deps: [], commits: opts.commits.length, source: "cloud" };
  for (const d of deps) {
    if (d.ai) { out.ai++; if (!d.reviewed) out.ai_unreviewed++; }
    if (d.checked && !d.exists) out.missing++;
    if (d.checked && d.exists && d.age_days >= 0 && d.age_days < YOUNG_DAYS) out.young++;
    if (suspicious(d.triage, d)) out.suspicious++;
  }
  out.deps = deps.sort((a, b) => rank(b) - rank(a)).slice(0, MAX_DEPS);
  return out;
}
